compiler/rustc_monomorphize/src/errors.rs

```diff
@@ -0,0 +1,74 @@
+use std::path::PathBuf;
+
+use rustc_errors::ErrorGuaranteed;
+use rustc_macros::SessionDiagnostic;
+use rustc_session::SessionDiagnostic;
+use rustc_span::Span;
+
+#[derive(SessionDiagnostic)]
+// FIXME: This should be #[fatal...], but that doesn't exist, yet...
+#[error(monomorphize::recursion_limit)]
+pub struct RecursionLimit {
+    #[primary_span]
+    pub span: Span,
+    pub shrunk: String,
+    #[note]
+    pub def_span: Span,
+    pub def_path_str: String,
+    #[note(monomorphize::written_to_path)]
+    pub was_written: Option<()>,
+    pub path: PathBuf,
+}
+
+#[derive(SessionDiagnostic)]
+// FIXME: This should be #[fatal...], but that doesn't exist, yet...
+#[error(monomorphize::type_length_limit)]
+#[help(monomorphize::consider_type_length_limit)]
+pub struct TypeLengthLimit {
+    #[primary_span]
+    pub span: Span,
+    pub shrunk: String,
+    #[note(monomorphize::written_to_path)]
+    pub was_written: Option<()>,
+    pub path: PathBuf,
+    pub type_length: usize,
+}
+
+#[derive(SessionDiagnostic)]
+// FIXME: This should be #[fatal...], but that doesn't exist, yet...
+#[error(monomorphize::fatal_error)]
+pub struct FatalError {
+    pub error_message: String,
+}
+
+#[derive(SessionDiagnostic)]
+// FIXME: This should be #[fatal...], but that doesn't exist, yet...
+#[error(monomorphize::fatal_error)]
+pub struct SpanFatalError {
+    #[primary_span]
+    pub span: Span,
+    pub error_message: String,
+}
+
+pub struct UnusedGenericParams {
+    pub span: Span,
+    pub param_spans: Vec<Span>,
+    pub param_names: Vec<String>,
+}
+
+impl SessionDiagnostic<'_> for UnusedGenericParams {
+    fn into_diagnostic(
+        self,
+        sess: &'_ rustc_session::parse::ParseSess,
+    ) -> rustc_errors::DiagnosticBuilder<'_, ErrorGuaranteed> {
+        let mut diag = sess.struct_err(rustc_errors::fluent::monomorphize::unused_generic_params);
```

Comment:
Hm, maybe we should expose some `sess.struct_$SOMETHING` (like `struct_diagnostic`?) that is generic over `EmissionGuarantee`, then make the `SessionDiagnostic` derive generic, i.e.

```
impl<'tcx> SessionDiagnostic for UnusedGenericParams {
  fn into_diagnostic<T: EmissionGuarantee>( .. ) -> DiagnosticBuilder<'tcx, T> {
    let mut diag = sess.struct_diagnostic(rustc_errors:..);
    ..
  }
}
```

cc @davidtwco who might have thoughts

Reply:
Like, ideally we shouldn't care what kind of `DiagnosticBuilder` (i.e. what kind of emission) we want when designing this struct. We should be able to do `tcx.emit_err(Foo);` or `tcx.emit_warn(Foo);` etc.

Reply:
I'd be happy with a change like this.